### ISSUE TITLE
`KeyError: 'logging'` when trying to upgrade to Kedro >= 0.18.8

### ISSUE THREAD

**Issue by astrojuanlu** [kedro-org/kedro#2721]
## Description

Pipelines that were working before trigger a `KeyError: 'logging'` when trying to upgrade to Kedro >= 0.18.8.

## Context

Reported internally by @matescharnitzky when trying to upgrade some pipelines.

See:

1. https://www.linen.dev/s/kedro/t/11976974/hi-everyone-we-re-in-the-process-of-relaxing-our-kedro-depen#a3a63b25-0748-4245-834f-5d043132e213
2. https://www.linen.dev/s/kedro/t/12656861/hi-team-currently-we-re-using-kedro-0-18-3-that-pins-pytest-#c94ed13f-fe4c-4dd9-944d-0c008aaf7861

## Steps to Reproduce

(Still working on a reproducer)

## Expected Result

Upgrading from 0.18.3 to 0.18.x should not trigger errors.

## Actual Result

```
Starting execution for corporate_banking_data_mocker
Starting sub process for the kedro run --pipeline corporate_banking_data_mocker 
After execution got the status code 1
╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│ /home/circleci/.pyenv/versions/3.9.17/bin/kedro:8 in <module>                │
│                                                                              │
│ /home/circleci/.pyenv/versions/3.9.17/lib/python3.9/site-packages/kedro/fram │
│ ework/cli/cli.py:211 in main                                                 │
│                                                                              │
│   208 │   """                                                                │
│   209 │   _init_plugins()                                                    │
│   210 │   cli_collection = KedroCLI(project_path=Path.cwd())                 │
│ ❱ 211 │   cli_collection()                                                   │
│   212                                                                        │
│                                                                              │
│ /home/circleci/.local/lib/python3.9/site-packages/click/core.py:1130 in      │
│ __call__                                                                     │
│                                                                              │
│ /home/circleci/.pyenv/versions/3.9.17/lib/python3.9/site-packages/kedro/fram │
│ ework/cli/cli.py:139 in main                                                 │
│                                                                              │
│   136 │   │   )                                                              │
│   137 │   │                                                                  │
│   138 │   │   try:                                                           │
│ ❱ 139 │   │   │   super().main(                                              │
│   140 │   │   │   │   args=args,                                             │
│   141 │   │   │   │   prog_name=prog_name,                                   │
│   142 │   │   │   │   complete_var=complete_var,                             │
│                                                                              │
│ /home/circleci/.local/lib/python3.9/site-packages/click/core.py:1055 in main │
│                                                                              │
│ /home/circleci/.local/lib/python3.9/site-packages/click/core.py:1657 in      │
│ invoke                                                                       │
│                                                                              │
│ /home/circleci/.local/lib/python3.9/site-packages/click/core.py:1404 in      │
│ invoke                                                                       │
│                                                                              │
│ /home/circleci/.local/lib/python3.9/site-packages/click/core.py:760 in       │
│ invoke                                                                       │
│                                                                              │
│ /home/circleci/.pyenv/versions/3.9.17/lib/python3.9/site-packages/kedro/fram │
│ ework/cli/project.py:456 in run                                              │
│                                                                              │
│   453 │   node_names = node_names + nodes_names                              │
│   454 │   load_version = {**load_version, **load_versions}                   │
│   455 │                                                                      │
│ ❱ 456 │   with KedroSession.create(                                          │
│   457 │   │   env=env, conf_source=conf_source, extra_params=params          │
│   458 │   ) as session:                                                      │
│   459 │   │   session.run(                                                   │
│                                                                              │
│ /home/circleci/.pyenv/versions/3.9.17/lib/python3.9/site-packages/kedro/fram │
│ ework/session/session.py:194 in create                                       │
│                                                                              │
│   191 │   │   session._store.update(session_data)                            │
│   192 │   │                                                                  │
│   193 │   │   # We need ConfigLoader and env to setup logging correctly      │
│ ❱ 194 │   │   session._setup_logging()                                       │
│   195 │   │   session_data.update(**_describe_git(session._project_path))    │
│   196 │   │   session._store.update(session_data)                            │
│   197                                                                        │
│                                                                              │
│ /home/circleci/.pyenv/versions/3.9.17/lib/python3.9/site-packages/kedro/fram │
│ ework/session/session.py:212 in _setup_logging                               │
│                                                                              │
│   209 │   def _setup_logging(self) -> None:                                  │
│   210 │   │   """Register logging specified in logging directory."""         │
│   211 │   │   try:                                                           │
│ ❱ 212 │   │   │   logging_config = self._get_logging_config()                │
│   213 │   │   except MissingConfigException:                                 │
│   214 │   │   │   self._logger.debug(                                        │
│   215 │   │   │   │   "No project logging configuration loaded; "            │
│                                                                              │
│ /home/circleci/.pyenv/versions/3.9.17/lib/python3.9/site-packages/kedro/fram │
│ ework/session/session.py:201 in _get_logging_config                          │
│                                                                              │
│   198 │   │   return session                                                 │
│   199 │                                                                      │
│   200 │   def _get_logging_config(self) -> dict[str, Any]:                   │
│ ❱ 201 │   │   logging_config = self._get_config_loader()["logging"]          │
│   202 │   │   # turn relative paths in logging config into absolute path     │
│   203 │   │   # before initialising loggers                                  │
│   204 │   │   logging_config = _convert_paths_to_absolute_posix(             │
│                                                                              │
│ /home/circleci/.pyenv/versions/3.9.17/lib/python3.9/collections/__init__.py: │
│ 1058 in __getitem__                                                          │
│                                                                              │
│   1055 │   │   │   return self.data[key]                                     │
│   1056 │   │   if hasattr(self.__class__, "__missing__"):                    │
│   1057 │   │   │   return self.__class__.__missing__(self, key)              │
│ ❱ 1058 │   │   raise KeyError(key)                                           │
│   1059 │                                                                     │
│   1060 │   def __setitem__(self, key, item):                                 │
│   1061 │   │   self.data[key] = item                                         │
╰───────────────���──────────────────────────────────────────────────────────────╯
KeyError: 'logging'
```

## Your Environment

* Kedro version used (`pip show kedro` or `kedro -V`): 0.18.10
* Python version used (`python -V`): 3.9.16
* Operating system and version: Linux

**Comment by astrojuanlu** [comment]
It's reported that 0.18.3 works, but 0.18.4 fails.

**Comment by astrojuanlu** [comment]
This happens when the user has a custom config loader that inherits from `AbstractConfigLoader`: in 0.18.4, internally Kedro started calling `config_loader[pattern]`, and so child classes broke.

Essentially this discussion: https://github.com/kedro-org/kedro/pull/1870#issuecomment-1262028818

Maybe a solution could be that `AbstractConfigLoader` implements a fallback `__getitem__` while issuing a `DeprecationWarning`, so at least the 0.18.x line is compatible.

**Comment by merelcht** [comment]
I've looked into this issue a bit more and my conclusion is as follows:
- This issue has been caused by the team directly inheriting from `AbstractConfigLoader` as well as importing methods that are meant to be private, indicated with a `_` at the start of the method name, from [`config.common.py` ](https://github.com/kedro-org/kedro/blob/main/kedro/config/common.py) (this is made on the assumption that they used this configloader: https://github.com/McK-Internal/c1_cltv_inigo/blob/44d5e3f68940d049bef17f6de8589a8398743f56/src/customerone_lib/common/jinja_templated_config_loader.py#L43, apologies this is in a private repo)
- Part of the problem here is that the abstraction for the config loader classes and the shared methods in `common.py` was done poorly and with just the `AbstractConfigLoader` it isn't possible to create a functioning config loader inheriting from it.
- The `OmegaConfigLoader` is not based on the code in `common.py` because of it's poor design. I hope that going forward with the `0.19.0` series we and our users will have a much better experience with the configloaders.
- I still consider this quite a niche problem since a large majority of our users either don't use custom config loader or inherit from `ConfigLoader` or `TemplatedConfigLoader` instead. 
- Therefore, I'll close this as won't fix.
